crypto/pkcs12/p12_attr.c

```diff
@@ -119,3 +119,8 @@ PKCS12_SAFEBAG_get0_attrs(const PKCS12_SAFEBAG *bag)
 {
     return bag->attrib;
 }
+
+void PKCS12_SAFEBAG_set_attrs(const STACK_OF(X509_ATTRIBUTE) *attrs, PKCS12_SAFEBAG *bag)
+{
+    bag->attrib = attrs;
```

Comment:
if `bag->attrib != attrs` then we should probably free `bag->attrib` first.

Reply:
That's the case in your code. But this PR is adding a new API function that could be called by any third party application. We do not know what the state of the bag is and whether the attributes are NULL or not. Consider this ridiculous example:

```` c
    STACK_OF(X509_ATTRIBUTE) *attr1 = sk_X509_ATTRIBUTE_new_null();
    STACK_OF(X509_ATTRIBUTE) *attr2 = sk_X509_ATTRIBUTE_new_null();
    PKCS12_SAFEBAG *bag = PKCS12_SAFEBAG_new();

    PKCS12_SAFEBAG_set_attrs(attr1, bag);
    /* "bag" now "owns" attr1 and takes over memory management for it. The application should not free it */

    /* ...do other stuff ... */
    /* Now we change our mind and want other attributes stored in the bag */
    PKCS12_SAFEBAG_set_attrs(attr2, bag);
    /* "bag" now "owns" attr2 and takes over memory management for it. The application should not free it */

    /* ...do something useful with bag... */

    /* Free the bag and all associated resources */
    PKCS12_SAFEBAG_free(bag);

    /* Memory leak! attr1 is leaked */
````

Reply:
Happy to change it but that example would look like an application level issue in them miss using the function, you could argue

Reply:
I don't agree that the example shows a misuse of the function. I would expect the example given to work correctly without a leak. In a "set0" function (which this is) the owning object takes over memory management for the thing being set. That means it has to free it, if it later discards it to be replaced by something else. This is common practice within the library.